src/io/inputs.jl

```diff
@@ -121,76 +122,65 @@ end
 
-    try
-        loaded = NamedDimsArray(data; zip(dim_names, dim_labels)...)
-    catch err
-        if isa(err, KeyError)
-            n_sites = size(data, 2)
-            @warn "Provided file $(data_fn) did not have the expected dimensions (one of: timesteps, reef_siteid, scenarios)."
-            if n_sites != nrow(site_data)
-                error(
-                    "Mismatch in number of sites ($(data_fn)). Expected $(nrow(site_data)), got $(n_sites)",
-                )
+        try
+            loaded_data = NamedDimsArray(data; zip(dim_names, dim_labels)...)
+        catch err
+            if isa(err, KeyError)
+                n_sites = size(data, 2)
+                @warn "Provided file $(data_fn) did not have the expected dimensions " *
+                    "(one of: timesteps, reef_siteid, scenarios)."
+                if n_sites != nrow(site_data)
+                    error(
+                        "Mismatch in number of sites ($(data_fn)). " *
+                        "Expected $(nrow(site_data)), got $(n_sites)",
+                    )
+                end
+            else
+                rethrow(err)
             end
-        else
-            rethrow(err)
         end
     end
 
-    return loaded
+    return loaded_data
 end
 
 """
-    _char_to_string(vals)::Vector{String}
+Return vector of labels for each array dimension. Important: Cannot trust indicated
+dimension metadata to get site labels, because this can be incorrect. Instead, match by
+number of sites.
+"""
+function _nc_dim_labels(
+    data_fn::String, data::Array{<:AbstractFloat}, nc_file::NetCDF.NcFile
+)::Vector{Union{UnitRange{Int64},Vector{String}}}
+    local sites_idx::Int64
 
-Convert character array entries in netCDFs to string.
+    has_reef_siteid_var = "reef_siteid" in keys(nc_file.vars)
+    sites = has_reef_siteid_var ? NetCDF.readvar(nc_file, "reef_siteid") : 1:size(data, 2)
 
-Some packages used to write out netCDFs do not yet support
-string values, and instead reverts to writing out character arrays.
-"""
-function _char_to_string(vals::AbstractVecOrMat)::Vector{String}
-    if vals isa Matrix
-        vals = map(x -> join(skipmissing(x)), eachcol(vals))
+    try
+        # This will be an issue if two or more dimensions have the same number of elements
+        # as the number of sites, but so far it hasn't happened...
+        sites_idx = first(findall(size(data) .== length(sites)))
+    catch err
+        error(
+            "Error loading $data_fn : could not determine number of locations." *
+            "Detected size: $(size(data)) | Known number of locations: $(length(sites))",
+        )
     end
 
-    # R's ncdf4 package does not yet support string values
-    # so strip the null terminator from the joined character array.
-    vals = replace.(vals, "\0" => "")
+    dim_labels = Union{UnitRange{Int64},Vector{String}}[1:n for n in size(data)]
+    dim_labels[sites_idx] = sites
 
-    return vals
+    return dim_labels
 end
```

Comment:
You sure character arrays are handled properly?

Reply:
Can you give me an example of what was the problem before? Is there a test for this (I can write one if you give me an example)?

Reply:
I have added a `::Vector{String}` to `sites` to be sure it is a vector of strings - and the tests are passing. Do you think this is enough?

`sites::Vector{String} = has_reef_siteid_var ? NetCDF.readvar(nc_file, "reef_siteid") : 1:size(data, 2)`

Reply:
No because it will just fail when reading a character array, rather than a string array. Wait a tick, I'm mocking up an example.

Reply:
I'm making an example netcdf but to give a concrete example, it's the difference between these two:

```julia

a = "a string"
b = ['c', 'h', 'a', 'r', 'a', 'c', 't', 'e', 'r']

typeof(a)
# String

typeof(b)
# Vector{Char} (alias for Array{Char, 1})

Reply:
Hum.. ok, I understand the difference (of course) but my question is: what was happening before that you wrote that function? NetCDF was reading each string as a vector of chars? Because that is definitely not happening now.

Reply:
I can confirm it doesn't automatically detect character arrays:

```julia
julia> ncread("./char_R_ncdf_test.nc", "reef_siteid")
14×3 Matrix{NetCDF.ASCIIChar}:
 'a'   's'   's'
 'b'   'o'   'o'
 'c'   'm'   'm'
 'd'   'e'   'e'
 '\0'  't'   't'
 '\0'  'h'   'h'
 '\0'  'i'   'i'
 '\0'  'n'   'n'
 '\0'  'g'   'g'
 '\0'  '\0'  ' '
 '\0'  '\0'  'e'
 '\0'  '\0'  'l'
 '\0'  '\0'  's'
 '\0'  '\0'  'e'
```

[char_R_ncdf_test.zip](https://github.com/open-AIMS/ADRIA.jl/files/13370916/char_R_ncdf_test.zip)

NetCDF.jl now seems to have `nc_char2string()` which handles it:

```julia
julia> nc_char2string(ncread("./char_R_ncdf_test.nc", "reef_siteid"))
3-element Vector{String}:
 "abcd"
 "something"
 "something else"
```

https://juliageo.org/NetCDF.jl/v0.5/strings.html#NC_CHAR-variables-1

Reply:
> Hum.. ok, I understand the difference (of course) but my question is: what was happening before that you wrote that function? NetCDF was reading each string as a vector of chars? Because that is definitely not happening now.

It's not happening now because you're reading in strings, not a character array. If we happen to get a netcdf which uses character arrays instead of strings, then the changed implementation will not work

Reply:
@ConnectedSystems I've added some tests together as well. Ready for another look

Reply:
@ConnectedSystems I've added the requested tests.